<a href="https://colab.research.google.com/github/frm1789/Data_words/blob/main/Data_analysis_SpaCy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## How many high frequency words are in a text?
In this case, I chose the book "Dinosaur at dark" from the well-known series of book, "The tree magic house".


1) Import Spacy and setting the language: English.

In [95]:
import spacy
nlp = spacy.load('en_core_web_sm')

2) Read the text and token it.

In [96]:
with open("text.txt", "r") as f:
  content = f.read()
  words = content.split()

doc = nlp(content)
words = [token.text for token in doc]

3) Clean non-alphabetic token


In [97]:
words = [word for word in words if word.isalpha()]
# len(words) -- 5522

4) Count how many unique words appear in the text

In [98]:
from collections import Counter
word_freq = Counter(words)
##print(word_freq)
print("Total quantity of unique words in text:", len(word_freq))

Total quantity of unique words in text: 1056


5) Create a unique list with the high frequency words from Kindergarten to Third Grade

In [99]:
hf_K = ['a', 'me', 'to', 'yes', 'big', 'I', 'go', 'in', 'cat', 'for', 'is', 'at', 'on', 'dog', 'he', 'the', 'you', 'like', 'up', 'she', 'mom', 'we', 'my', 'with', 'this', 'dad', 'it', 'by',
        'said', 'look', 'can', 'no', 'love', 'play', 'went', 'see', 'am', 'do', 'was', 'and']
hf_F = ['after', 'have', 'please', 'all', 'her', 'saw', 'an', 'here', 'should', 'are', 'him', 'so', 'as', 'his', 'some', 'be', 'I’m', 'thank', 'because', 'if', 'that', 'but', 'into', 'them', 
        'came', 'just', 'then', 'come', 'know', 'they', 'could', 'little', 'there', 'day', 'make', 'us', 'did', 'many', 'very', 'end', 'new', 'want', 'from', 'not', 'were', 'get', 'of', 'what',
        'goes', 'one', 'when', 'going', 'or', 'where', 'good', 'our', 'who', 'had', 'out', 'will', 'has', 'over', 'would','your' ]
hf_S = ['after', 'have', 'please', 'all', 'her', 'saw', 'an', 'here', 'should', 'are', 'him', 'so', 'as', 'his', 'some', 'be', 'I’m', 'thank', 'because', 'if', 'that', 'but', 'into', 'them', 
        'came', 'just', 'then', 'come', 'know', 'they', 'could', 'little', 'there', 'day', 'make', 'us', 'did', 'many', 'very', 'end', 'new', 'want', 'from', 'not', 'were', 'get', 'of', 'what',
        'goes', 'one', 'when', 'going', 'or', 'where', 'good', 'our', 'who', 'had', 'out', 'will', 'has', 'over', 'would','your' ]
hf_T = ['also', 'animal', 'April', 'asked', 'August', 'books', 'brother', 'call', 'care', 'caught', 'caught', 'children', 'city', 'December', 'drink', 'eight', 'eleven',
        'ever', 'father', 'favorite', 'February', 'found', 'Friday', 'friend', 'give', 'also', 'animal', 'April', 'asked', 'August', 'books', 'brother', 'call', 'care',
        'caught', 'children', 'city', 'December', 'drink', 'eight', 'eleven', 'ever', 'father', 'favorite', 'February', 'found', 'Friday', 'friend', 'give', 'great', 
        'also' 'animal' 'April', 'asked', 'August', 'books', 'brother', 'call', 'care', 'caught', 'caught', 'children', 'city', 'December', 'drink', 'eight', 'eleven',
        'ever', 'father', 'favorite', 'February', 'found', 'Friday', 'friend', 'give', 'great', 'gym', 'half', 'hour', 'hurt', 'January', 'July', 'June', 'kind', 'line', 
        'mail', 'March', 'May', 'maybe', 'Michigan', 'might', 'Monday', 'mother', 'must', 'night', 'nine', 'November', 'October', 'old', 'other', 'people', 'pretty', 'questions',
        'rain', 'rest', 'same', 'Saturday', 'September', 'seven', 'sister', 'snow', 'something', 'stay', 'street', 'such', 'Sunday', 'sure', 'talk', 'teacher', 'tell', 
        'thing', 'Thursday', 'too', 'Tuesday', 'twelve', 'Wednesday', 'phone']


In [100]:
combined_list = hf_K + hf_F + hf_S + hf_T
unique_list = list(set(combined_list))
unique_list.sort()
#print(unique_list)
print("Total list of word that children learn as high frequency word from Kindergarten to Third Grade: ", len(unique_list))

Total list of word that children learn as high frequency word from Kindergarten to Third Grade:  181


6) Count the ocurrence of each of the HF words in the text 

In [101]:
from collections import Counter
filter_word = Counter(word_freq)
counts = {word: filter_word[word] for word in combined_list}
#print(counts)

In [102]:
total_count = 0
for key, value in counts.items():
    total_count += value
print("\nTotal quantity of high frequency words in the text: ", total_count)
print("\nTotal words of text", len(words))
percentage = total_count/len(words)*100
#print("\nTotal words that belongs of the category high frequency words - 1 Grade: ", total)
print("\nTotal percentage of words:", percentage)


Total quantity of high frequency words in the text:  2102

Total words of text 5522

Total percentage of words: 38.06591814559942


### Conclusion:

A child in third grade can read 38% of the text.

Third grade is the year when kids go from 'learning to read' to 'reading to learn'. A 38% of the text is not enough.

Reading proficiently by the end of third grade is a crucial marker in a child's educational development. Failure to read proficiently is linked to higher rates of school dropout.

So, high frequency words and balanced literature didn't work to generate a proficient reader. 
